docs/nima/testing.adoc

```diff
@@ -0,0 +1,397 @@
+        ws.endpoint("/testWs", WS_LISTENER);
+    }
+
+    @Test
+    void testWsEndpoint() { <3>
+        WS_LISTENER.reset();
+
+        ClientSideListener clientListener = new ClientSideListener();
+        wsClient.connect("/testWs", clientListener);    <4>
+
+        assertThat(clientListener.message, is("ws"));   <5>
+    }
+}
+----
+<1> Declare `WsClient` and later inject it in the constructor;
+<2> Using `@SetUpRoute` create WebSocket routing and assign a serverside listener;
+<3> Test the WebSocket endpoint using regular `@Test` annotation;
+<4> Create and assign clientside listener;
+<5> Check if the received message is correct;
+
+
+FIXME: not sure we need this, maybe as popup?
+
+WebSocket `ClientSideListener` is a helper class implementing `WsListener`, is very simple and looks this way:
+
+.`ClientSideListener` helper class.
+[source,java]
+----
+private static class ClientSideListener implements WsListener {
+    private final CountDownLatch cdl = new CountDownLatch(1);
+    private String message;
+    private volatile Throwable throwable;
+
+    @Override
+    public void onOpen(WsSession session) { <1>
+        session.send("hello", true);
+    }
+
+    @Override
+    public void onMessage(WsSession session, String text, boolean last) {   <2>
+        this.message = text;
+        session.close(WsCloseCodes.NORMAL_CLOSE, "End");
+    }
+
+    @Override
+    public void onClose(WsSession session, int status, String reason) { <3>
+        cdl.countDown();
+    }
+
+    @Override
+    public void onError(WsSession session, Throwable t) {   <4>
+        this.throwable = t;
+        cdl.countDown();
+    }
+}
+----
+<1> Send "Hello", when a connection is opened;
+<2> Save the message when received and close the connection;
+<3> Close the connection;
+<4> React on an error;
+
+WebSocket `ClientSideListener` is also a helper class implementing `WsListener` and it is very straightforward:
+
+.`ServerSideListener` helper class.
+[source,java]
+----
+private static class ServerSideListener implements WsListener {
+    boolean opened;
+    boolean closed;
+    String message;
+
+    @Override
+    public void onMessage(WsSession session, String text, boolean last) {   <1>
+        message = text;
+        session.send("ws", true);
+    }
+
+    @Override
+    public void onClose(WsSession session, int status, String reason) {     <2>
+        closed = true;
+    }
+
+    @Override
+    public void onOpen(WsSession session) {     <3>
+        opened = true;
+    }
+
+    void reset() {          <4>
+        opened = false;
+        closed = false;
+        message = null;
+    }
+}
+----
+<1> Send "ws" on a received message;
+<2> Called when the connection is called;
+<3> Called on connection opened;
+<4> Used to reset the state;
+
+If only routing tests are required without real port opening, the testing class should be annotated with `@RoutingTest`. Instead of `WsClient` – `DirectWsClient` should be used.
```

Comment:
My suggestion:
The testing class should be annotated with `@RoutingTest` only if routing tests are required without real port opening. Instead of `WsClient`, use `DirectWsClient`.

Reply:
Ok, Thank you!